File: [I]stack-queue/[I]stack-queue/155-min-stack.py
class MinStack:

    def __init__(self):
        self.stack = []
        self.min_num = None

    def push(self, val: int) -> None:
        if self.min_num == None:
            self.stack.append(0)
            self.min_num = val
        elif val < self.min_num:
            self.stack.append(val - self.min_num)
            self.min_num = val
        else:
            self.stack.append(val - self.min_num)

    def pop(self) -> None:
        diff = self.stack.pop()
        if len(self.stack) == 0:
            self.min_num = None
        elif diff < 0:
            self.min_num += abs(diff)

    def top(self) -> int:
        diff = self.stack[- 1]
        if diff < 0:
            return self.min_num
        return self.min_num + diff

    def getMin(self) -> int:
        return self.min_num

# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

# time O(1)
# space O(n)
# using stack and queue and implement stack/queue and one stack

File: [I]stack-queue/[I]stack-queue/155-min-stack.java
class MinStack {
    ArrayDeque<Long> stack;
    long min;

    public MinStack() {
        stack = new ArrayDeque<>();
        min = 0;
    }
    
    public void push(int val) {
        if (stack.isEmpty()) {
            min = val;
            stack.push(0L);
        } else {
            long diff = val - min;
            if (diff < 0) {
                min = val;
                stack.push(diff);
            } else {
                stack.push(diff);
            }
        }
    }
    
    public void pop() {
        Long diff = stack.peek();
        if (diff < 0) {
            min += Math.abs(diff);
            stack.pop();
        } else {
            stack.pop();
        }
        if (stack.isEmpty()) {
            min = 0;
        }
    }
    
    public int top() {
        Long diff = stack.peek();
        if (diff < 0) {
            return (int) min;
        }
        return (int) (min + diff);
    }
    
    public int getMin() {
        return (int) min;
    }
}

/**
* Your MinStack object will be instantiated and called as such:
* MinStack obj = new MinStack();
* obj.push(val);
* obj.pop();
* int param_3 = obj.top();
* int param_4 = obj.getMin();
*/

// time O(1)
// space O(n)
// using stack and queue and implement stack/queue and one stack